auth/crammd5/mechanism.go

```diff
@@ -0,0 +1,42 @@
+package crammd5
+
+import (
+	"crypto/hmac"
+	"crypto/md5"
+	"encoding/hex"
+	"errors"
+	"io"
+
+	log "github.com/golang/glog"
+)
+
+var (
+	illegalStateErr = errors.New("illegal mechanism state")
+)
+
+type mechanism struct {
+	username string
+	secret   []byte
+}
+
+type stepFunc func(m *mechanism, data []byte) (stepFunc, []byte, error)
+
+// algorithm lifted from wikipedia: http://en.wikipedia.org/wiki/CRAM-MD5
+// except that the SASL mechanism used by Mesos doesn't leverage base64 encoding
+func challengeResponse(m *mechanism, data []byte) (stepFunc, []byte, error) {
+	decoded := string(data)
+	log.V(4).Infof("challenge(decoded): %s", decoded) // for deep debugging only
+
+	hash := hmac.New(md5.New, m.secret)
+	if _, err := io.WriteString(hash, decoded); err != nil {
+		return illegalState, nil, err
+	}
+
+	codes := hex.EncodeToString(hash.Sum(nil))
+	msg := m.username + " " + codes
+	return nil, []byte(msg), nil
+}
+
+func illegalState(m *mechanism, data []byte) (stepFunc, []byte, error) {
+	return illegalState, nil, illegalStateErr
```

Comment:
Good question. I debated this for a bit myself, then failed to document my thinking: the intent is that `StepFunc` implementations should never return a `nil` StepFunc result. This makes the logic in the SASL authticatee simpler: step functions are never `nil`. Step functions that end up an error state (for example, some decoding logic fails...) should return a StepFunc that represents an error state. Some mechanisms may be able to recover from such. Not CRAMMD5. I'll add a TODO for docs around StepFunc.